File: lib/goals.py
"""Financial goal planning — SIP required, feasibility, allocation hints."""

from __future__ import annotations

from dataclasses import dataclass


@dataclass
class GoalPlan:
    name: str
    target: float
    years: int
    priority: int
    required_monthly_sip: float
    recommended_asset: str
    on_track: bool
    gap_monthly: float


def monthly_sip_for_goal(
    target: float,
    years: int,
    annual_return: float = 0.12,
    current_saved: float = 0.0,
) -> float:
    """PMT: monthly SIP needed to reach target."""
    months = max(years * 12, 1)
    r = annual_return / 12 / 100 if annual_return > 1 else annual_return / 12
    fv_existing = current_saved * ((1 + r) ** months)
    remaining = max(0, target - fv_existing)
    if r == 0:
        return remaining / months
    return remaining * r / (((1 + r) ** months - 1) * (1 + r))


def asset_for_horizon(years: int) -> str:
    if years <= 3:
        return "Debt / FD / Liquid funds"
    if years <= 7:
        return "Hybrid (40% equity max)"
    return "Equity-heavy (index + diversified)"


def plan_goal(
    name: str,
    target: float,
    years: int,
    priority: int,
    current_saved: float,
    allocated_sip: float,
    annual_return: float = 12.0,
) -> GoalPlan:
    required = monthly_sip_for_goal(target, years, annual_return, current_saved)
    gap = max(0, required - allocated_sip)
    return GoalPlan(
        name=name,
        target=target,
        years=years,
        priority=priority,
        required_monthly_sip=round(required, 0),
        recommended_asset=asset_for_horizon(years),
        on_track=allocated_sip >= required * 0.95,
        gap_monthly=round(gap, 0),
    )

File: web/src/lib/goals.ts
import type { GoalPlan } from "./types";

export function monthlySipForGoal(
  target: number,
  years: number,
  annualReturn = 0.12,
  currentSaved = 0
): number {
  const months = Math.max(years * 12, 1);
  const r = annualReturn > 1 ? annualReturn / 12 / 100 : annualReturn / 12;
  const fvExisting = currentSaved * Math.pow(1 + r, months);
  const remaining = Math.max(0, target - fvExisting);
  if (r === 0) return remaining / months;
  return (remaining * r) / ((Math.pow(1 + r, months) - 1) * (1 + r));
}

export function assetForHorizon(years: number): string {
  if (years <= 3) return "Debt / FD / Liquid funds";
  if (years <= 7) return "Hybrid (40% equity max)";
  return "Equity-heavy (index + diversified)";
}

export function planGoal(
  name: string,
  target: number,
  years: number,
  priority: number,
  currentSaved: number,
  allocatedSip: number,
  annualReturn = 12
): GoalPlan {
  const required = monthlySipForGoal(target, years, annualReturn, currentSaved);
  const gap = Math.max(0, required - allocatedSip);
  return {
    name,
    target,
    years,
    priority,
    required_monthly_sip: Math.round(required),
    recommended_asset: assetForHorizon(years),
    on_track: allocatedSip >= required * 0.95,
    gap_monthly: Math.round(gap),
  };
}
